rfc/src/rfcs/0002-function-stubbing.md

```diff
@@ -0,0 +1,416 @@
+At this stage of the compiler, names have been fully resolved, and there is no problem with swapping in the body of a function defined in one crate for a function defined in another.
+
+The major downside with the MIR-to-MIR transformation is that it does not appear to be possible to stub types at that stage (there is no way to change the definition of a type through the MIR).
+Thus, our proposed approach will not be a fully general stubbing solution.
+However, it is technically feasible and relatively clean, and provides benefits over having no stubbing at all (as can be seen in the examples in the first part of this document).
+
+Furthermore, it can be used as part of a portfolio of stubbing approaches, where users stub local types using conditional compilation (see Alternative #1), and Kani provides a modified version of the standard library with verification-friendly versions of types like `std::vec::Vec`.
+
+### Alternative #1: Conditional compilation
+
+In this baseline alternative, we do not provide any stubbing mechanism at all.
+Instead, users can effectively stub local code (functions, methods, and types) using conditional compilation.
+For example, they could specify using `#[cfg(kani)]` to turn off the original definition and turn on the replacement definition when Kani is running, similarly to the ghost state approach taken in the [Tokio Bytes proof](https://model-checking.github.io/kani-verifier-blog/2022/08/17/using-the-kani-rust-verifier-on-tokio-bytes.html).
+
+The disadvantage with this approach is that it does not provide any way to stub external code, which is one of the main motivations of our proposed approach.
+
+### Alternative #2: Source-to-source transformation
+
+In this alternative, we rewrite the source code before it even gets to the compiler.
+The advantage with this approach is that it is very flexible, allowing us to stub functions, methods, and types, either by directly replacing them, or appending their replacements and injecting appropriate conditional compilation guards.
+
+The downside with this approach is that it requires all source code to be available.
+It also might be difficult to inject code in a way that names are correctly resolved (e.g., if the replacement code comes from a different crate).
+Also, source code is difficult to work with, and includes things like unexpanded macros.
+
+On the last two points, we might be able to take advantage of an existing source analysis platform like `rust-analyzer` (which has facilities like structural search replace), but this would add more (potentially fragile) dependencies to Kani.
+
+### Alternative #3: AST-to-AST or HIR-to-HIR transformation
+
+In this alternative, we implement stubbing by rewriting the AST or [High-Level IR (HIR)](https://rustc-dev-guide.rust-lang.org/hir.html) of the program.
```

Comment:
Is there documentation for AST as there is for HIR? If so, I suggest linking it here.